### ISSUE TITLE
Asciidoctor warns about incorrect list item index when list numbering starts from arbitrary number

### ISSUE THREAD

**Issue by skalee** [asciidoctor/asciidoctor#3252]
Following AsciiDoc document is pretty self-explanatory. For readability, rendered version is in this gist: https://gist.github.com/skalee/69f868601aa1cb36aa5a79f7aa939fc3.

```adoc
= Probably a bug

== About this list

1.  This list has many items, which are split across several sections.
2.  List items in document include index number.  That is, they are formatted
    like:
+
--------------------------------------------------------------------------------
1. This list has many items, which are split across several sections.
--------------------------------------------------------------------------------
+
not like:
+
--------------------------------------------------------------------------------
. This list has many items, which are split across several sections.
--------------------------------------------------------------------------------

== Why it is split?

[start=3]
3.  There are use cases in which starting from higher number is advocated.
4.  For example, when many lists (from AsciiDoctor's point of view) make sense
    as a whole (from reader's point of view).
5.  I mean, software installation procedure could be organized in sections like:
    "Hardware requirements", "Software prerequisites", "Where to download from",
    "Compiling C++ sources", etc.
6.  This bug report is also kinda valid use case.

== It renders correctly

[start=7]
7.  Actually, this list is rendered correctly.
8.  AsciiDoc supports `[start=n]` annotation which allows to continue list from
    some higher number.
9.  And yes, this technique is used in this very document.

== So, what's wrong with it?

[start=10]
10.  Warnings!
11.  It says:
+
--------------------------------------------------------------------------------
asciidoctor: WARNING: Wrong.adoc: line 21: list item index: expected 1, got 3
asciidoctor: WARNING: Wrong.adoc: line 22: list item index: expected 2, got 4
asciidoctor: WARNING: Wrong.adoc: line 23: list item index: expected 3, got 5
asciidoctor: WARNING: Wrong.adoc: line 24: list item index: expected 4, got 6
asciidoctor: WARNING: Wrong.adoc: line 29: list item index: expected 1, got 7
asciidoctor: WARNING: Wrong.adoc: line 30: list item index: expected 2, got 8
asciidoctor: WARNING: Wrong.adoc: line 31: list item index: expected 3, got 9
asciidoctor: WARNING: Wrong.adoc: line 36: list item index: expected 1, got 10
asciidoctor: WARNING: Wrong.adoc: line 37: list item index: expected 2, got 11
asciidoctor: WARNING: Wrong.adoc: line 52: list item index: expected 3, got 12
asciidoctor: WARNING: Wrong.adoc: line 57: list item index: expected 1, got 13
--------------------------------------------------------------------------------
12. Which is definitely incorrect.

== Are you up to date?

[start=13]
13. Yes, `asciidoctor --version` prints following:
+
--------------------------------------------------------------------------------
Asciidoctor 2.0.6 [https://asciidoctor.org]
Runtime Environment (ruby 2.6.1p33 (2019-01-30 revision 66950) [x86_64-darwin17]) (lc:UTF-8 fs:UTF-8 in:UTF-8 ex:UTF-8)
--------------------------------------------------------------------------------
```

**Comment by mojavelinux** [comment]
You should not be using explicit numbers in ordered lists. There's just no need for it. That's only supported for backwards compatibility with AsciiDoc Python. (Even the warning is emulating AsciiDoc Python's behavior).

**Comment by rockyallen** [comment]
But there are use cases for it, and the current warning is quite annoying and not helpful.

**Use case 1:  Lists that represent steps in a procedure** 
Step numbering often has to be continuous across subheadings, hence [start=]. The steps are numbered explicitly so that it is easy to reference steps by their number. eg
````
[start=5]
5. Remove the bolt from the port bracket
6. Remove the bracket from the assembly
7. Repeat step 5 to step 6 for the starboard bracket.
````
Without explicit numbering there is a huge danger of getting the referenced step numbers wrong, and no way to automatically check when it happens.
I would much prefer to write:
````
. [[step-bracket]]Remove the bolt from the port bracket
. [[step-assembly]]Remove the bracket from the assembly
. Repeat <<step-bracket>> to <<step-assembly>> for the starboard bracket.
````
But currently this only works in Docbook. HTML and pdf create a hyperlink that does not show the step number. I know I can over-ride the displayed text, but that has exactly the same danger that I was trying to avoid.
An accurate warning message here would detect real errors; currently they are lost in the warning-storm.

**Use case 2: A 2-column legend for an image with callouts**
````
[[fig-pump]]
image::pump.jpg[]
[role=legend]
|===
|[start=1]
1. Bracket
2. Assembly
|[start=3]
3. Bolt
4. Washer
|===
````
I use a list because it gives me exactly the formatting I need, very concisely.
The items are numbered explicitly so that I can write 'remove bolt (3) and lift bracket (1) from assembly (2)'.
In this case even 'improved' cross-references as above would not help, because the numbers must match the callout numbers embedded in the preceding image.
Perhaps call outs would be more semantic than a list, but I would I lose the elegant formating, and would they work with images?

**Comment by mojavelinux** [comment]
This seems to me like a duplicate of #2218, then. There are still unaddressed questions in that issue. Until we solve them, we wouldn't be able to move forward with this anyway. I'm open to considering it, though.

> But currently this only works in Docbook. HTML and pdf create a hyperlink that does not show the step number. 

I was not aware the DocBook stylesheets had this behavior. This seems to me like a separate enhancement we should consider. It's closely related to #2721 (and also related to xrefstyle).

**Comment by rockyallen** [comment]
Agreed that this issue will not be as important when those other issues are fixed, but this one is still a bug, and fixing it would make life more comfortable now. I will raise an enhancement request for the other.

**Comment by mojavelinux** [comment]
I have to look deeper, but solving #2218 may be required to address this request.

Since the current behavior matches AsciiDoc Python, this is not a bug but rather a compliance enhancement. I implemented it exactly to specification. We're just changing the specification.

**Comment by mojavelinux** [comment]
I'll tentatively schedule for 2.1.0. That may change if I determine a fix for #2218 is required.

**Comment by rmannibucau** [comment]
Hello,

Any hope it gets solved? This kind of issue is quite common when inserting code snippets between points.

Thanks

**Comment by mojavelinux** [comment]
I won't be working on it anytime soon. I still stand by my recommendation of not using explicit numbers, even though I do understand and support the argument for using them in certain circumstances.

**Comment by rmannibucau** [comment]
@mojavelinux is there a built-in alternative to do [first point][source or table block][second point][source block]? If not, not sure how to follow the recommendation :(.

**Comment by mojavelinux** [comment]
It sounds like you're talking about a list with continuations. Is there something about what you tried that doesn't work? If you provide an example, I can help you fix it.

**Comment by rmannibucau** [comment]
Sure, here is the example (simplified to not have a ton of text for the demo):


    To setup the library you must follow these steps:
    
    
    1. Create a `$APPLICATION_BASE/conf/conf.d` folder if it does not exist
    2. Create a `server.properties` file in this folder with this content:
    
    [source,properties]
    ----
    disabled = false
    port = 1234
    ----
    
    [start=3]
    3. Set the following properties in `$APPLICATION_BASE/conf/system.properties`:
    
    [source,properties]
    ----
    serialization.format = JSON
    serialization.whitelist = com.acme.*
    ----

**Comment by mojavelinux** [comment]
It's unnecessary to break up the list. You can use this:

```
To setup the library you must follow these steps:

1. Create a `$APPLICATION_BASE/conf/conf.d` folder if it does not exist
2. Create a `server.properties` file in this folder with this content:
+
[source,properties]
----
disabled = false
port = 1234
----

3. Set the following properties in `$APPLICATION_BASE/conf/system.properties`:
+
[source,properties]
----
serialization.format = JSON
serialization.whitelist = com.acme.*
----
```

(I'd still argue that the explicit numbers aren't necessary).

**Comment by rmannibucau** [comment]
Thanks Dan, works perfectly for my cases.

**Comment by mojavelinux** [comment]
Awesome! Glad to hear it.

**Comment by geekq** [comment]
Thanks for the hint of using _list continuations_ and using `.` instead of explicit numbers!

For other people looking for further details: the documentation for list continuations and a complex example can be found at https://docs.asciidoctor.org/asciidoc/latest/lists/ordered/ > "When not to use the start attribute" and
https://docs.asciidoctor.org/asciidoc/latest/lists/continuation/#list-continuation

It looks like a list can not continue over a subheader like `=== next steps`, even with a `+`. For that I used `[start=7]`

**Comment by mojavelinux** [comment]
What I'm considering allowing is the following:

```
5. five
6. six
```

Thus far, we've classified this to be a syntax error as it's usually an indication that a list continuation was not used properly. We would lose that check. On the other hand, it's an intuitive way to start a list as a specified ordinal when using explicit numbers. Since we can't have it both ways, I'd like to know what people's opinions are on this trade-off.

**Comment by mojavelinux** [comment]
The challenge with honoring the start attribute is that the attributes are assigned to the block after the block is parsed. However, it's easy enough to forward the start attribute on to the parse_list method (as we do with other attributes) so that it can be consulted when building the list. Pretty straightforward change, in the end.
